py/farm_ng/core/events_file_writer.py

```diff
@@ -135,10 +139,50 @@ def _increment_file_idx(self) -> None:
         """Increment the file index."""
         self._file_idx += 1
 
+    @property
+    def header_msgs(self) -> list[tuple[Event, bytes]]:
+        """Return the list of header messages.
+        Returns:
+            list[tuple[Event, bytes]]: List of header messages.
+        """
+        return self._header_msgs
+
+    def add_header_msg(self, event: Event, payload: bytes, write: bool = False) -> None:
+        """Add a header message, and optionally writes it to the file.
+        NOTE: Writing to file will fail if the file is not open.
+
+        Args:
+            event: Event to write.
+            payload: Payload to write.
+            write: If True, write the header message to the file. Defaults to False.
+        """
+        if not isinstance(event, Event):
+            error_msg = f"event must be Event, not {type(event)}"
+            raise TypeError(error_msg)
+        if not isinstance(payload, bytes):
+            error_msg = f"payload must be bytes, not {type(payload)}"
+            raise TypeError(error_msg)
+        self._header_msgs.append((event, payload))
```

Comment:
Here I would suggest making header msgs be a dict based on the uri, so that for any header message uri we only support one message so that it can't grow in an unbounded way

Reply:
Yeah great idea. This simplifies the risky handling you noted above

Reply:
OK @ethanrublee current intended behavior:

- `RecorderService` allows for overwriting header fields with a matching `Uri` (dict key is `uri_to_string(Uri)`)
- `EventsFileWriter` does NOT allow for overwriting fields with a matching `Uri` (same key)

This way once you send the `start` command to the `RecorderService`, the headers the `RecorderService` has on hand are added to the log file and all subsequent file splits. If any headers with a new `Uri` are received while logging, they are written to the current file split immediately and are added to the header in all future file splits. Any headers on an existing Uri are ignored by the `eventsFileWriter`. This prevents header mismatch between log file splits (with the exception of added header messages).


After that logging session is complete (send `stop` command or another `start` and `EventsFileWriter` is closed), the next `EventsFileWriter` logging session will use any updated header messages sent to the `RecorderService` during the previous logging session or since its conclusion.


Last thing is for the latched topics that we use as headers (e.g. `/oak0/calibration`). The topic + service names are defined in the config for the `RecorderService`, but the `RecorderService` never actually sees that message or stores it with the manually defined headers (like the key-value pair `metadata` in the camera app). Rather the `EventServiceRecorder` receives it from the latching publisher when it is created by the `RecorderService` `start` command. It passes the message to the `EventsFileWriter` to add to the header messages for this logging session. We could pass these up to the `RecorderService`to maintain as well if we want, but it is not currently doing that.

Please let me know:
1. If this is reasonable behavior
2. If we should pass the latched messages up to the `RecorderService` to maintain